Agrego TDA Mazo y TDA Carta. Usan nodo.h

guillehoff opened the pull request:
TDA Mazo, con TDA Carta y TDA Nodo. El mazo es una cola, con nodos que tienen ademas del puntero una carta, solo con el atributo descripcion. (No se si habia q escribir algo asi)

guillehoff committed:
Add files via upload

NLGS2907 commented on Carta.h:
```diff
@@ -0,0 +1,35 @@
+#ifndef CARTA_H_
+#define CARTA_H_
+
+#include<iostream>
```
```c++
#include <iostream>
```
¿Esta línea es necesaria?
No estoy viendo que uses `std::cout` ni nada parecido.

guillehoff replied:
Si "#include <iostream>" no es necesario, mala costumbre de ponerlo automatico.
Perfecto ahora paso todo a solo .h y uso nodo.h. 
Voy a probar Mazo.h como template y vuelvo a hacer pull request.

NLGS2907 replied:
ok, no es necesario hacer un nuevo pull request, si no modificar este basta (creo que no podés tener 2 a la vez así que cuando lo hagas va a editar este solo)

NLGS2907 replied:
Aún veo el `#include <iostream>` ahí xDDD
Igual no es vital, después se arregla cuando hagamos correción masiva de documentación.

guillehoff replied:
lol ahi lo saque

NLGS2907 reviewed (changes requested):
Los archivos `Nodo.h` y `Nodo.cpp` son incompatibles y sobreescribirían los que y hice. (En realidad no, porque el mío esta todo en minúscula y el tuyo empieza con mayúscula, así que son dos diferentes) Pero igual recomiendo que uses los que ya tenés en tu rama que codeé yo y que están hechos con templates.
Al hacerlo en templates, probablemente debas borrar también `Carta.cpp` y 'Mazo.cpp', debido a que hacemos los métodos también dentro del ***.h** .

guillehoff committed:
Elimino y subo Carta.h hecho con Template

guillehoff committed:
Elimino y subo Carta.h con Template

guillehoff committed:
Subo Carta.h (template) contiene definiciones

guillehoff committed:
Voy a utilizar nodo.h

guillehoff committed:
Voy a utilizar nodo.h

guillehoff committed:
Subo Mazo.h con uso de nodo.h y template

No se si en los  metodos de TDA nodo y Carta, que estan dentro de los metodos de Mazo hay que ponerle T en argumento.
Porque en la definicon de los metodos esta T

Ejemplo: 

Linea 93-> Nodo<T>* nuevoNodo = new Nodo<T>(descripcionCarta); 
Linea 99-> this->fondo->setProx(nuevoNodo); 

¿descripcionCarta y nuevoNodo necesitan T adelante?

guillehoff committed:
Subi Mazo.h que reemplaza este

guillehoff committed:
Saque #include<iostream>

guillehoff committed:
Cambie a minuscula

guillehoff committed:
Cambie a minuscula nombre de file

NLGS2907 commented on carta.h:
```diff
@@ -0,0 +1,55 @@
+template <class T>
+Carta<T>::Carta(T descripcion)  {
+
+	this->descripcion = descripcion;
```
Este archivo es impecable, pero solo una consulta:
la razón de que yo no ponga `this->atributo` en mis codigos es por casos como los de esta linea.
Aunque no son lo mismo `this->informacion` e `informacion` podrían ser malentendidos como el mismo dato por tener nombres tan similares, normalmente hago algo del tipo:

```c++
template <class T>
Carta<T>::Carta(T DescripcionCarta) {

      descripcion = descripcionCarta;
}
```
o similar. Otra cosa es que el nombre `descripcion` da a entender que el mismo va a ser una info en forma de *<string>*, y que no necesariamente es cierto. Hacerlo con templates significa que la info que guarda tu TDA podría ser *<int>*, *<char>* o incluso otro TDA. Por eso, ¿Podría recomendar quizás nombre más genericos como `datoCarta` o `infoCarta`, para que tampoco quede colgados y muy ambiguos si fueran `dato` e `info`?
Lo mismo conque el tipo sea `T`. No hay nada de malo, pero como es medio genérico, me gusta ponerle algo más literal como `tipoCarta`, aún si después alarga el código jajaja.

De nuevo, todo esto es yo siendo exquisito, y que puede ser medio ~~paja~~ flojera tener que cambiar cada método ahora, pero puede dar problemas como los que explico más abajo.

guillehoff replied:
Gracias por la recomendacion que si es con template tiene sentido que sea datoCarta en vez de descripcion!

NLGS2907 commented on mazo.h:
```diff
@@ -0,0 +1,152 @@
+#ifndef MAZO_H_
+#define MAZO_H_
+
+#include"Nodo.h"
```
haciendo pruebas, descubrí que los `#include` no son *case sensitive*, es decir que tanto el archivo se llama así como la línea ésta si dice `"nodo.h"` como si dice `"Nodo.h"`, `"NODO.h"` o `"NoDo.h"` funca igual. No es problema mayor esto, es lo que digo :)

De todas formas en el futuro haré que todos sigamos en lo posible una convención. Por supuesto están invitado a protestar si están en desacuerdo.

NLGS2907 commented on mazo.h:
```diff
@@ -0,0 +1,152 @@
+template <class T>
+void Mazo<T>::acolar(T descripcionCarta){ // NO PONGO Nodo<T>* siguiente=NULL Porque tiene parametro por defecto
+
+	Nodo<T>* nuevoNodo = new Nodo<T>(descripcionCarta);
```
Como puse arriba, `descripcionCarta` es medio confuso. Pero igual, el problema es que estás creando el nodo, si entendí bien, con la descripción de la Carta como dato, en vez de la carta en sí. ¿Me explico? Si `Carta.descripcion` resultaba ser un dato de tipo `std::string`.  Estarías guardando en este caso un nodo que guarda un string, no el TDA Carta.

Todo esto claro, si no tenemos en cuenta que `descripcionCarta` sea en realidad un TDA Carta y no un string o int ni nada. `descripcionCarta` es un *nombre* a fin de cuentas nada más.

Me parece que te parece confuso y vas a mezclar después los métodos propios de una Cola con los que serían exclusivos de nuestro TDA Mazo. Yo sugiero crear otro TDA más llamado `Cola.h` donde poner todos estos métodos que ponés, obviamente usando nombre más genéricos como `dato` o `info` como en `listaLigada.h`, ya que podría ser que después quieras armar una cola con un dato que no sea un TDA Carta. Al estar integrado en `Mazo.h`, sólo estás guardando un tipo de dato que es un TDA Carta.
Después si acaso en `Mazo.h` podés hacer `Mazo mazitoGuay = Cola<Carta>()` para inicializar el mazo y además así separás modularizando los métodos y atributos que sólo se usarían en un TDA Mazo pero no en una Cola genérica en dos archivos diferentes.

NLGS2907 commented on mazo.h:
```diff
@@ -0,0 +1,152 @@
+	Carta<T>* carta = NULL;
+
+    if (this->estaVacia()) {
+        throw std::string("El mazo esta vacio");
```
Esto, al menos a mí, lanza una excepción e interrumpe la ejecución del programa pero no imprime por pantalla el mensaje `"El mazo está vacío"`, como probablemente estás queriendo hacer. La escapada que yo hice en mis TDAs es incluir un `#include <iostream>` (ahora sí xD) arriba de todo y luego acá poner:
```c++

    if (this->estaVacia()) {
       std::cout << "El mazo está vacío" << std::endl;
       throw /* [Acá bien podría ir cualquier tipo de dato, pero en este método no encontré ninguno] */;
```

NLGS2907 commented on mazo.h:
```diff
@@ -0,0 +1,152 @@
+	Nodo<T>* nuevoNodo = new Nodo<T>(descripcionCarta);
+
+	if (this->estaVacia()) {
+		this->frente = nuevoNodo;
```
acá no modificás en nada `this->fondo`, de modo que sigue apuntando a `NULL`. Ahora bien, es medio peliagudo, porque si lo pensás bien, al haber sólo un elemento en la Cola el único nodo presente debería ser `frente` y `fondo` al mismo tiempo, es decir que estos dos atributos deberían de apuntar al mismo nodo.

guillehoff replied:
Entendi que cuando estaVacia es TRUE, fondo = nuevoNodo, para que deje de apuntar a NULL. 
Asi que agregue this->fondo = nuevoNodo; ahi.

NLGS2907 commented on mazo.h:
```diff
@@ -0,0 +1,152 @@
+		this->frente = nuevoNodo;
+
+	} else {
+		this->fondo->setProx(nuevoNodo);
```
¿Pensás la cola como `(1) frente <- x <- x <- x <- fondo` o como `(2) frente -> x -> x -> x -> frondo`?
Porque parece que acá estás agregando el nuevo dato al siguiente del `this->fondo`, si lo estás pensando como `(1)` te quedaría `frente <- x <- x <- x <- nuevoNodo <- fondo` y no acolarse correctamente.

Aún si lo pensaste como en `(2)`, como nunca los uniste en el constructor ni tampoco tocaste `this->fondo` como dije arriba, nunca establecés la relación `frente -> fondo` y los dos Nodos quedan aislados. Si seguís apilando de esta manera te quedás con: 
`frente -> NULL     -|-     fondo`
`frente -> NULL     -|-     viejoFondo -> nuevoFondo`
`frente -> NULL     -|-     viejoFondo -> nuevoFondo -> nuevoNuevísimoFondo`
y nunca estarán conectadas en la misma Cola.

De todas formas igual, recomiendo pensarlo como `(1)` porque está bueno pensar el fondo, de donde se agrega como el "principio"; y el frente, de donde se sacan cosas, como el "final". La operación no cambiaría mucho:
```c++
       this->fondo = nuevoNodo->setProx(this->fondo)
```

NLGS2907 commented on mazo.h:
```diff
@@ -0,0 +1,152 @@
+    }
+
+
+	if (!this->estaVacia()) {
```
este `if` es redundante porque estás repitiendo la operacion `this->estaVacia` por segunda vez. Simplemente conectalo a un `else` en el `if` de arriba, o bien, no pongas ningun `if` acá para empezar, porque de todas formas si la condición allá arriba se cumple, el programa rompe con el `throw`. Si llegás hasta este punto la Cola siempre va a estar **no** vacía.

NLGS2907 commented on mazo.h:
```diff
@@ -0,0 +1,152 @@
+		Nodo<T>* frenteAnterior = this->frente;
+		this->frente = frenteAnterior->prox();
+
+	    if (this->frente == NULL) {
```
solo por acostubrarse a usar tus propios métodos, probá igual con `this->estaVacia()`

NLGS2907 commented on mazo.h:
```diff
@@ -0,0 +1,152 @@
+	    }
+
+	 /* devuelve el elemento y libera los recursos asociados */
+	    carta = frenteAnterior->dato(); //dato() METODO de nodo.h devuelve dato de nodo
```
esto es cosa mía, pero ¿Cambia en algo si en vez de al principio de todo el método, a `Carta<T>* carta = NULL;` lo declarás recién acá? Total si explota con el throw que pusiste, estarías creando una variable que después no se usaría.

guillehoff replied:
Lo puse arriba pq intento declarar nuevas var al principio, pero tiene sentido lo que decis. 
Lo modifique.

NLGS2907 replied:
acá no tanto, pero esto suele ser un tema ya especialmente con memoria dinámica, a veces se me escapa a mí y hago excepciones que dejan riesgo de que quede colgada.

NLGS2907 commented on mazo.h:
```diff
@@ -0,0 +1,152 @@
+        throw std::string("El mazo esta vacio");
+    }
+
+	if (!this->estaVacia()) {
```
ídem de errores similares de más arriba...

NLGS2907 commented on mazo.h:
```diff
@@ -0,0 +1,152 @@
+
+	if (!this->estaVacia()) {
+
+		carta = this->frente->dato();
```
te podrías ahorrar crear una variable nueva si simplemente pusieras `reurn this->frente->dato()` al final y ya todos pa' casa.

NLGS2907 reviewed (changes requested):
@guillehoff, lo siento en el alma pero hay que rehacerlo bastante (yo te ayudo si querés, vos llamame cuando quieras):

Mi único comentario final es que repienses un TDA Cola olvidándote de usarlo como un mazo.
Pensá que sólo querés desarrollar un TDA Cola que contenga cualquier tipo de dato, eso te va a llevar a métodos más generales que te van a ayudar si después querés usar el TDA para otra cosa. `Mazo.h` sólo sería una implementación de ese TDA Cola, de la misma forma que mi `tablero.h` solo es una implementación de `listaLigada.h`.

Tal cual está tiene errores que no me dejaron compilarlo, después consultanos a cualquiera sobre los cambios y te digo sin problema, a mi me encanta programar y no me gustaría que se les haga embole a otros, sobre todo cuando todos estamos en esta carrera que justamente nos prepara para esto jeje. Como dije cualquier cosa no dudes en avisarme o protestarme sobre algo, yo tampoco soy sabelotodo ;)




**Protip:** *Sugiero agregar un método `imprimir()` que saque por consola el estado actual de la Cola, creeme que ayuda una barbaridad en debug*

guillehoff committed:
Arregle lo recomendado

Puse datoCarta como nombre del atributo  y al parametro datoCartaNuevo porque no se meocurrio una mejor cosa. Si alguien se ilumina y tiene una idea de algo mas representativo cambio el nombre.

guillehoff committed:
Hice cambios, dados en comentarios. 

Puse todo esto asi me acuerdo q hice. Pq faltan cosas.
agregue #include<iostream>, para throw. 
97 - this->fondo = nuevoNodo; (cuando esta vacia apuntan los dos punteros al unico nodo)
123 - Carta<T>* carta = NULL; (movi declaracion variable, dsps del throw)
144 - return this->frente->dato();  (ahorrar crear una variable nueva )

guillehoff committed:
TDA cola (el que dio el curso) y TDA carta2 este no es Template

guillehoff committed:
No sirve este mazo.h

guillehoff committed:
cola.h con metodos de nodo.h de este TP

cola.h del curso tenia metodos del nodo.h tambien del curso asi que los cambie por los equivalentes de nodo.h de este TP

guillehoff committed:
Agregue metodo imprimir, es como el de Franco en listaLigada

Use el mismo metodo que Franco para imprimir. 
Si se les ocurre algo mejor lo cambiamos.

guillehoff committed:
agregue mazo.h 

TDA Mazo con uso de cola.h y carta2.h

guillehoff committed:
Agregue constructor sin parametro

guillehoff committed:
Add files via upload

guillehoff committed:
genera de 0 a 3 ahora en lugar de 1 a 3

NLGS2907 commented on rand.cpp:
```diff
@@ -0,0 +1,35 @@
+int main() {
+
+	/* inicializa la generación de números aleatorios */
+		srand(time(NULL));
```
Esto te funciona? A mí me tira error

guillehoff replied:
En eclipse me funciona.
No sabia q era de c, y no de c++. 
No habia visto que habia que usar < random>

guillehoff replied:
Ahora aprendo <random > y lo cambio. 
Asi usamos solo librerias de c++

NLGS2907 replied:
ok dale, el problema es que no sé si < random> es del estándar C++98, por lo que si tenés problemas con eso avisanos y le decimos a los profes a ver qué hacemos.

guillehoff replied:
Si, < random > no es del estándar C++98. Ahi lo arregle para que funcione con el estándar C++98, estaba con otro estándar por eso me funcionaba.

NLGS2907 commented on rand.cpp:
```diff
@@ -0,0 +1,35 @@
+		for (int i = 0; i < 10; i++) {
+
+			/* entero aleatorio entre 0 y 2147483647 */
+			int aleatorio = rand();
```
Esto tambien me tira error

NLGS2907 reviewed (commented):
Vine a ver porque me daba curiosidad pero me pareció muy raro ver las funciones srand() y rand(), que por lo que vi no son de < random>.
Esas funciones provienen de <stdlib.h>, que diría es más de C.
Por las dudas igual te pregunto; esto a mí no me quiere funcar jajaja

NLGS2907 commented:
Busqué y no encontré forma. Retiro lo que dije en los comentarios, y ya me decís vos cómo hiciste, antes de mandarme cagadas

guillehoff committed:
agregue int aleatorio2

guillehoff committed:
Uso de libreria <random>

Libreria <random> de c++.

guillehoff committed:
carta.h con Template no tiene uso

guillehoff committed:
cambie nombred e carta2 a carta

guillehoff committed:
Usa rand() con <cstdlib>

guillehoff committed:
Update carta.h

guillehoff committed:
Update mazo.h

guillehoff committed:
agregue menu.h

guillehoff committed:
Agregue algunas postcondiciones

guillehoff committed:
no util

guillehoff committed:
innecesario

NLGS2907 reviewed (approved):
Yo recién voy a probar todo esto cuando termine el menú, pero por lo pronto, me funca bien. APROBADO PAPU

guillehoff committed:
Elimino para subir correspondiente .cpp y .h

guillehoff committed:
Elimino para subir correspondiente .cpp y .h

guillehoff committed:
Added mazo.h mazo.cpp carta.h y carta.cpp

NLGS2907 commented on cola.h:
```diff
@@ -0,0 +1,149 @@
+
+#include"nodo.h"
+#include<string>
+
```
Para el metodo imprimir() estás usando `std::cout` y `std::endl`, pero para eso te falta incluir `#include <iostream >` por acá

NLGS2907 reviewed (commented):
Me da error el constructor del mazo, no sé por qué

EDIT: ya lo descubrí: te falto inicializar en el constructor el mazo, en específico la línea `this->mazo = new Cola<Carta*>()`

NLGS2907 commented on mazo.cpp:
```diff
@@ -0,0 +1,123 @@
+/* ----- DEFINICIONES ----- */
+
+Mazo::Mazo(int cantidadDeCartasJuegaDoble, int cantidadDeCartasBloquearTurno, int cantidadDeCartasAgarrarCincoFichas, int cantidadDeCartasEliminarMazoJugadorSiguiente) {
+
```
Como dije en algún comentario, acá falta la línea `this->mazo = new Cola<Carta*>();` para inicializar el atributo `this-mazo`, pero el caso es que igual este algoritmo, tal y como está, es muy costoso e ineficiente.
Por suerte Nico descubrió que lo que jodía tanto era el crear memoria dinámica para después borrarla si no se acolaba al mazo.
Acá te propongo una versión modificada que funciona perfecto:
```c++
Mazo::Mazo(int cantidadDeCartasJuegaDoble, int cantidadDeCartasBloquearTurno, int cantidadDeCartasAgarrarCincoFichas, int cantidadDeCartasEliminarMazoJugadorSiguiente) {

        mazo = new Cola<Carta*>();

	    int cantidadTotal = cantidadDeCartasJuegaDoble + cantidadDeCartasBloquearTurno + cantidadDeCartasAgarrarCincoFichas + cantidadDeCartasEliminarMazoJugadorSiguiente;

	    int contador1 = 0;
	    int contador2 = 0;
	    int contador3 = 0;
	    int contador4 = 0;

	    int cantidadDeCartas = 0;

	    while (cantidadDeCartas < cantidadTotal) {

	        CartaEspecial tipoAleatorio = generarNumeroRandom();
	        if (tipoAleatorio == JuegaDoble) {

	            if (contador1 < cantidadDeCartasJuegaDoble) {

	                mazo->acolar(new Carta(tipoAleatorio));
	                contador1++;
	                cantidadDeCartas++;

	            }
	        } else if (tipoAleatorio == BloquearTurno) {

	            if (contador2 < cantidadDeCartasBloquearTurno) {

	                mazo->acolar(new Carta(tipoAleatorio));
	                contador2++;
	                cantidadDeCartas++;

	            }
	        } else if (tipoAleatorio == AgarrarCincoFichas) {

	            if (contador3 < cantidadDeCartasAgarrarCincoFichas) {

	                mazo->acolar(new Carta(tipoAleatorio));
	                contador3++;
	                cantidadDeCartas++;

	            }

	        } else if (tipoAleatorio == EliminarMazoJugadorSiguiente) {

	            if (contador3 < cantidadDeCartasEliminarMazoJugadorSiguiente) {

	                mazo->acolar(new Carta(tipoAleatorio));
	                contador4++;
	                cantidadDeCartas++;

	            }
	        }
	    }

}
```

También como comentario extra, noté que tenés un método `acolarCarta()` en el TDA Mazo. Si lo usás queda más piola porque no se ve el atributo interno `this->mazo`:
```c++
Mazo::Mazo(int cantidadDeCartasJuegaDoble, int cantidadDeCartasBloquearTurno, int cantidadDeCartasAgarrarCincoFichas, int cantidadDeCartasEliminarMazoJugadorSiguiente) {

        mazo = new Cola<Carta*>();

	    int cantidadTotal = cantidadDeCartasJuegaDoble + cantidadDeCartasBloquearTurno + cantidadDeCartasAgarrarCincoFichas + cantidadDeCartasEliminarMazoJugadorSiguiente;

	    int contador1 = 0;
	    int contador2 = 0;
	    int contador3 = 0;
	    int contador4 = 0;

	    int cantidadDeCartas = 0;

	    while (cantidadDeCartas < cantidadTotal) {

	        CartaEspecial tipoAleatorio = generarNumeroRandom();
	        if (tipoAleatorio == JuegaDoble) {

	            if (contador1 < cantidadDeCartasJuegaDoble) {

	                this->acolarCarta(new Carta(tipoAleatorio));
	                contador1++;
	                cantidadDeCartas++;

	            }
	        } else if (tipoAleatorio == BloquearTurno) {

	            if (contador2 < cantidadDeCartasBloquearTurno) {

	                this->acolarCarta(new Carta(tipoAleatorio));
	                contador2++;
	                cantidadDeCartas++;

	            }
	        } else if (tipoAleatorio == AgarrarCincoFichas) {

	            if (contador3 < cantidadDeCartasAgarrarCincoFichas) {

	                this->acolarCarta(new Carta(tipoAleatorio));
	                contador3++;
	                cantidadDeCartas++;

	            }

	        } else if (tipoAleatorio == EliminarMazoJugadorSiguiente) {

	            if (contador3 < cantidadDeCartasEliminarMazoJugadorSiguiente) {

	                this->acolarCarta(new Carta(tipoAleatorio));
	                contador4++;
	                cantidadDeCartas++;

	            }
	        }
	    }

}
```

NLGS2907 reviewed (commented):
Ahí propuse una solución al error del constructor del tablero

guillehoff reviewed (commented):
Ahora subo corregido.
Muchas gracias.

guillehoff committed:
Agregue <iostream> para uso de cout, etc

guillehoff committed:
Correcion de constructor de Mazo.
